### DISCUSSION TITLE
How to return Callable containing generic paramter?

### DISCUSSION THREAD

**Discussion by tamuhey** [microsoft/pyright#1491] [Q&A] [answered]
The below function `f` returns `Callable[[T],T]`, but I don't know how to annotate this funciton:

```python
from typing import Callable, TypeVar

T = TypeVar("T")
F = Callable[[T], T]

def f() -> F: # reportMissingTypeArgument
    def g(x: T) -> T:
        ...
    return g

g = f()
x = g("foo") # report general type issues
```

If I omit the return type signature of `f`, it works:

```python
def f(): # omit return type
    def g(x: T) -> T:
        ...
    return g

g = f()
x = g("foo") # No error
```

But I want to annotate `f` precisely. How to do that?

**Comment by tamuhey** [comment, marked as answer]
I noticed that simply adding `Callable[[T], T]` works, but aliasing `F = Callable[[T],T]` doesn't work.

**Reply by erictraut** [reply]
`F` is a generic type alias, but you are not providing a type argument for it.

The best solution is the one you've already discovered, to avoid the use of a type alias in this case and use `Callable[[T], T]` as the return type annotation.

You should be able to use the type alias as long as you provide a type argument for it (e.g. `F[T]`), but I just noticed that there's a bug in the logic that specializes a generic type alias that includes a Callable if the type argument is another type variable. I'll work on a fix for this.

**Reply by erictraut** [reply]
Incidentally, if you had enabled the `reportMissingTypeArgument` diagnostic rule (or enabled "strict" type checking mode), pyright would have emitted an error for the missing type argument for `F`.
